bodhi/server/consumers/masher.py

```diff
@@ -388,16 +389,22 @@ def __init__(self, release, request, updates, agent,
         self.success = False
         self.devnull = None
         self._startyear = None
+        self.max_concur_sem = max_concur_sem
 
     def run(self):
         """Run the thread by managing a db transaction and calling work()."""
+        self.log.info('Grabbing semaphore')
+        self.max_concur_sem.acquire()
+        self.log.info('Acquired semaphore, starting')
         try:
             with self.db_factory() as session:
                 self.db = session
                 self.work()
                 self.db = None
         except Exception:
             self.log.exception('MasherThread failed. Transaction rolled back.')
+        self.max_concur_sem.release()
+        self.log.info('Released semaphore')
```

Comment:
It might be good to put the above two lines in a ```finally``` block, even though the ```Exception``` handler currently catches all. That way it's a bit more future proof in case the ```Exception``` handler is altered. It can also guard against the call to ```log.exception()``` raising an ```Exception``` of its own.